src/base/IoManip.h

```diff
@@ -220,5 +220,39 @@ operator <<(std::ostream &os, const AsList<Container> &manipulator)
 template <typename Container>
 inline auto asList(const Container &c) { return AsList<Container>(c); }
 
+
+/** output the constructed argument at most once per lifetime of the
+ * AtMostOnce object. The parameter must provide operator<<(std::ostream&)
+ * A const reference to the printed object is stored; it's up to the
+ * caller to manage object lifetime.
+ */
+template <class T>
+class AtMostOnce
+{
+    public:
+    explicit AtMostOnce(const T& t) : toPrint(t), printed(false) {}
+
+    std::ostream& print(std::ostream& os) {
+        if (!printed) {
+            os << toPrint;
+            printed = true;
+        }
+        return os;
+    }
+
+    private:
+    const T& toPrint;
+    bool printed = false;
+};
+
+/// a helper to ease AtMostOnce object creation
+template <class T>
+inline auto atMostOnce(const T &t) { return AtMostOnce<T>(t); }
+
+template <class T>
+inline std::ostream& operator<<(std::ostream& os, AtMostOnce<T>& a) {
+    return a.print(os);
+}
```

Comment:
How exactly is the **once** property supposed to be guaranteed when every call to this generates a new temporary instance of `class AtMostOnce` ?

Reply:
Yes it does. See TestIoManip::testAtMostOnce for intended use patterns.
Unless I'm missing something

Reply:
The target pattern is:

```c++
  auto once = atMostOnce("some heading\n");
  ostream << once << "some text\n";
  ostream << once << "some more text" << anObject << '\n';
```

Reply:
> Amos: How exactly is the once property supposed to be guaranteed when every call to this generates a new temporary

> Francesco: Yes it does.

You are both right. When "this" is atMostOnce() function, every call generates a new instance. When "this" is "<<" operator, every call does not generate a new instance. The latter is how this new feature is expected to be used (as illustrated by Francesco).

We need to research whether the helper function is needed in this (and similar) cases. If it is not needed, we should not be adding it here (and in similar cases). I know such helper functions are needed in _some_ cases, but I do not remember what exactly defines that need/those cases. I will research this when I start reviewing this PR, but if you can find the answer, please do!

Reply:
AFAIK these helper functions are used to facilitate template argument deduction, so that
```
auto a1 = atMostOnce("const char *");
and 
std::string s1("std::string");
auto a2 = atMostOnce(s1);
```
work out of the box

Reply:
Actually, after a quick experiment, it turns out to not be needed.
Removing.

Reply:
In general, such helper functions are not needed in some cases and are needed in others. It would be good to know the actual difference instead of adding/removing code based on possibly incomplete tests.

Reply:
A [Microsoft devblog article](https://devblogs.microsoft.com/cppblog/how-to-use-class-template-argument-deduction/) can help. 

> Class Template Argument Deduction (CTAD) is a C++17 Core Language feature that reduces code verbosity [...]
> C++98 through C++14 performed template argument deduction for function templates [..]
> However, class templates didn’t benefit from these rules. If you wanted to construct a std::pair from two ints, you had to say std::pair<int, int> p(11, 22);, despite the fact that the compiler already knows that the types of 11 and 22 are int